sql: during asynchronous add column, UPDATE uses new schema while its data source uses the old

```
Build Tag:   beta-20170105
Build Time:  2017/01/05 17:47:53
Platform:    linux amd64
Go Version:  go1.7.3
C Compiler:  gcc 4.9.2
```

I have a table:
```
CREATE TABLE goods_descriptions_info (
    id STRING(36) NOT NULL,
    md5 STRING(32) NOT NULL,
    length INT NOT NULL,
    CONSTRAINT "primary" PRIMARY KEY (id),
    INDEX goods_descriptions_info_md5_idx (md5),
    INDEX goods_descriptions_info_length_idx (length),
    FAMILY "primary" (id, md5, length)
)    
```

I was filled it with 6M records and ran:
`ALTER TABLE goods_descriptions_info ADD client_id int NOT NULL DEFAULT 0;`

Before it finished I was ran another query:
`UPDATE goods_descriptions_info SET length = 27000 WHERE md5 = 'fb54aa6272b31628abff07c48ec7c7db';`

And server is shutdown. Here is log:

```
I170109 09:09:26.426250 190 server/status/runtime.go:228  runtime stats: 6.1 GiB RSS, 151 goroutines, 24 MiB/85 MiB/141 MiB GO alloc/idle/total, 4.7 GiB/5.8 GiB CGO alloc/total, 292451.56cgo/sec, 0.86/0.03 %(u/s)time, 0.01 %gc (71x)
panic: invalid var index 3 (columns: 3) [recovered]
        panic: UPDATE goods_descriptions_info SET length = 27000 WHERE md5 = 'fb54aa6272b31628abff07c48ec7c7db';: invalid var index 3 (columns: 3) [recovered]
        panic: UPDATE goods_descriptions_info SET length = 27000 WHERE md5 = 'fb54aa6272b31628abff07c48ec7c7db';: invalid var index 3 (columns: 3)

goroutine 548132 [running]:
panic(0x187dc60, 0xc422cb4930)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc42021def0)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x6e
panic(0x187dc60, 0xc422cb4930)
        /usr/local/go/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements.func1(0xc422b6118b, 0x61)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:455 +0x149
panic(0x18370e0, 0xc422cb48c0)
        /usr/local/go/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql/parser.(*IndexedVarHelper).checkIndex(0xc423a5d350, 0x3)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/indexed_vars.go:141 +0x177
github.com/cockroachdb/cockroach/pkg/sql/parser.(*IndexedVarHelper).IndexedVarUsed(0xc423a5d350, 0x3, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/indexed_vars.go:166 +0x35
github.com/cockroachdb/cockroach/pkg/sql.setNeededColumns(0x2639320, 0xc423a5d000, 0xc422cb4878, 0x4, 0x4)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/needed_columns.go:90 +0xfcf
github.com/cockroachdb/cockroach/pkg/sql.setNeededColumns(0x26390a0, 0xc424069980, 0xc422cb4878, 0x4, 0x4)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/needed_columns.go:59 +0x3b6
github.com/cockroachdb/cockroach/pkg/sql.setNeededColumns(0x26393a0, 0xc4215d8fc0, 0xc422cb4870, 0x5, 0x5)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/needed_columns.go:136 +0xb64
github.com/cockroachdb/cockroach/pkg/sql.setNeededColumns(0x2639620, 0xc420a94000, 0x2b45510, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/needed_columns.go:186 +0x1500
github.com/cockroachdb/cockroach/pkg/sql.(*planner).optimizePlan(0xc422666170, 0x2639620, 0xc420a94000, 0x2b45510, 0x0, 0x0, 0x1, 0x2639620, 0xc420a94000, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/optimize.go:41 +0xf0
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makePlan(0xc422666170, 0x2630da0, 0xc421403ef0, 0xebc0001, 0x2af3860, 0xc423295e90, 0x991e82, 0xed0054a4d)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:210 +0x12b
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmt(0xc4201aa160, 0x2630da0, 0xc421403ef0, 0xc422666170, 0xc423290001, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1313 +0xa3
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn(0xc4201aa160, 0x2630da0, 0xc421403ef0, 0xc422666170, 0x101, 0xc422666068, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1113 +0x2ab
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtsInCurrentTxn(0xc4201aa160, 0xc422cb4100, 0x1, 0x1, 0xc422666170, 0xc422666068, 0x101, 0x0, 0x6, 0x22cb4140, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:850 +0x7db
github.com/cockroachdb/cockroach/pkg/sql.runTxnAttempt(0xc4201aa160, 0xc422666170, 0x1, 0xc422666068, 0xc422cb4160, 0xc422cb4100, 0x1, 0x1, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:768 +0xf8
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest.func2(0xc420d13760, 0xc422cb4160, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:621 +0x2b7
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec(0xc420d13760, 0xc422560101, 0xc420308780, 0xc422568d90, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:520 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc4201aa160, 0xc422666000, 0xc422b6118b, 0x61, 0x0, 0x0, 0x0, 0x0, 0xc420783300)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:643 +0x5b0
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc4201aa160, 0xc422666000, 0xc422b6118b, 0x61, 0x0, 0x0, 0x0, 0x0, 0xc422b61100)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:461 +0x11f
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).executeStatements(0xc4232ed000, 0x7f728dd71840, 0xc422f11650, 0xc422b6118b, 0x61, 0x0, 0x0, 0x0, 0x0, 0x735c01, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:794 +0x94
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc4232ed000, 0x7f728dd71840, 0xc422f11650, 0xc4232ed028, 0x66, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:468 +0xb3
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc4232ed000, 0x7f728dd71840, 0xc422f11650, 0x5400, 0xc4202122f8, 0x7f728dd71840, 0xc422f11620, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:395 +0xa05
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc4202121c0, 0x7f728dd71840, 0xc422f11620, 0x2638260, 0xc4202ea9a0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:309 +0x999
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func8.1(0x2638260, 0xc4202ea9a0)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:522 +0x106
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc42021def0, 0xc420138158, 0x2638260, 0xc4202ea9a0, 0xc42072fa60)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:136 +0x95
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
        /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:138 +0x277

```

--- a/pkg/sql/index_join.go
+++ b/pkg/sql/index_join.go
@@ -125,7 +125,7 @@ func (p *planner) makeIndexJoin(
 	// Create a new scanNode that will be used with the primary index.
 	table := p.Scan()
 	table.desc = origScan.desc
-	table.initDescDefaults(publicColumns)
+	table.initDescDefaults(origScan.scanVisibility)
 	table.initOrdering(0)
 	table.disableBatchLimit()
 

--- a/pkg/sql/scan.go
+++ b/pkg/sql/scan.go
@@ -83,7 +83,7 @@ type scanNode struct {
 	limitHint          int64
 	limitSoft          bool
 	disableBatchLimits bool
-
+	scanVisibility     scanVisibility
 	// This struct must be allocated on the heap and its location stay
 	// stable after construction because it implements
 	// IndexedVarContainer and the IndexedVar objects in sub-expressions
@@ -273,6 +273,7 @@ func (n *scanNode) initTable(
 
 // Initializes the column structures.
 func (n *scanNode) initDescDefaults(scanVisibility scanVisibility) {
+	n.scanVisibility = scanVisibility
 	n.index = &n.desc.PrimaryIndex
 	n.cols = make([]sqlbase.ColumnDescriptor, 0, len(n.desc.Columns)+len(n.desc.Mutations))
 	switch scanVisibility {
@@ -282,7 +283,11 @@ func (n *scanNode) initDescDefaults(scanVisibility scanVisibility) {
 		n.cols = append(n.cols, n.desc.Columns...)
 		for _, mutation := range n.desc.Mutations {
 			if c := mutation.GetColumn(); c != nil {
-				n.cols = append(n.cols, *c)
+				col := *c
+				// Even if the column is non-nullable it can be null in the
+				// middle of a schema change.
+				col.Nullable = true
+				n.cols = append(n.cols, col)
 			}
 		}
 	}

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -1552,3 +1552,71 @@ CREATE TABLE t.test (
 		t.Fatalf("read the wrong number of rows: e = %d, v = %d", eCount, count)
 	}
 }
+
+// Test an UPDATE using a primary and a secondary index in the middle
+// of a column backfill.
+func TestUpdateDuringColumnBackfill(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	backfillNotification := make(chan bool)
+	continueBackfillNotification := make(chan bool)
+	params, _ := createTestServerParams()
+	params.Knobs = base.TestingKnobs{
+		SQLSchemaChanger: &csql.SchemaChangerTestingKnobs{
+			RunBeforeBackfillChunk: func(sp roachpb.Span) error {
+				if backfillNotification != nil {
+					// Close channel to notify that the schema change has
+					// been queued and the backfill has started.
+					close(backfillNotification)
+					backfillNotification = nil
+					<-continueBackfillNotification
+				}
+				return nil
+			},
+			AsyncExecNotification: asyncSchemaChangerDisabled,
+		},
+	}
+	server, sqlDB, _ := serverutils.StartServer(t, params)
+	defer server.Stopper().Stop()
+
+	if _, err := sqlDB.Exec(`
+CREATE DATABASE t;
+CREATE TABLE t.test (
+    k INT NOT NULL,
+    v INT NOT NULL,
+    length INT NOT NULL,
+    CONSTRAINT "primary" PRIMARY KEY (k),
+    INDEX v_idx (v),
+    FAMILY "primary" (k, v, length)
+);
+INSERT INTO t.test (k, v, length) VALUES (0, 1, 1);
+`); err != nil {
+		t.Fatal(err)
+	}
+
+	// Run the column schema change in a separate goroutine.
+	notification := backfillNotification
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		if _, err := sqlDB.Exec(`ALTER TABLE t.test ADD id int NOT NULL DEFAULT 0;`); err != nil {
+			t.Error(err)
+		}
+		wg.Done()
+	}()
+
+	<-notification
+
+	// UPDATE the row using the secondary index.
+	if _, err := sqlDB.Exec(`UPDATE t.test SET length = 27000 WHERE v = 1`); err != nil {
+		t.Error(err)
+	}
+
+	// UPDATE the row using the primary index.
+	if _, err := sqlDB.Exec(`UPDATE t.test SET length = 27001 WHERE k = 0`); err != nil {
+		t.Error(err)
+	}
+
+	close(continueBackfillNotification)
+
+	wg.Wait()
+}

